sklearn/metrics/classification.py

```diff
@@ -1415,15 +1422,33 @@ def balanced_accuracy_score(y_true, y_pred, sample_weight=None):
 
     """
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
+    check_consistent_length(y_true, y_pred, sample_weight)
 
-    if y_type != 'binary':
+    if y_type.startswith('multilabel'):
         raise ValueError('Balanced accuracy is only meaningful '
-                         'for binary classification problems.')
-    # simply wrap the ``recall_score`` function
-    return recall_score(y_true, y_pred,
-                        pos_label=None,
-                        average='macro',
-                        sample_weight=sample_weight)
+                         'for binary and multiclass problems.')
+
+    unique_classes, encoded_y_true = np.unique(y_true, return_inverse=True)
+
+    missing_from_true = np.setdiff1d(y_pred, unique_classes)
+    if len(missing_from_true):
+        warnings.warn('y_pred contains classes not in y_true')
+
+    # This emulates compute_sample_weight('balanced', y_true),
+    # but accounts for sample weight when balancing classes.
+    n_classes = len(unique_classes)
+    class_weight = 1 / np.bincount(encoded_y_true, weights=sample_weight,
+                                   minlength=n_classes)
+    if sample_weight is None:
+        sample_weight = 1
+    sample_weight = class_weight.take(encoded_y_true) * sample_weight
```

Comment:
What is the reason to apply `sample_weight` a second time. I thought it was already taken into account when computing the `class_weight`. Which paper should I check for references?

Reply:
I don't think weighted balanced accuracy is reported anywhere, but:
* the PR's implementation matches the incumbed
* the implementation matches the invariance tests for weighting in `metrics.tests.test_common` that are really pretty good, if I must say so myself as the architect...

Generally when we have a value for `class_weight` as well as `sample_weight`, we weight the samples per the `class_weight` (i.e. `class_weight.take(y)`), and then we multiply by each sample's weight. Exactly what's happening here. However, currently our handling of `class_weight='balanced'` counts the *number* not the *total weight* of samples in each class, then assigns each the reciprocal as each sample's weight. I initially used that in this implementation, and was not surprised to find that it failed the tests: repetition of samples was no longer equivalent to integer weights. So here we use the total weight (not the cardinality) in determining the class weight, reciprocate that, but still assign each sample its weight so that we can correctly calculate the weighted confusion matrix.

Which makes me think: we should be able to implement this even more simply from the confusion matrix... I'll play with that another time soon.